### ISSUE TITLE
ComboboxControl: SuggestionsList loses focus on DOWN key press

### ISSUE THREAD

**Issue by mahnunchik** [WordPress/gutenberg#38548]
### Description

Whole block including `ComboboxControl` loses focus on `DOWN` key press.

### Step-by-step reproduction instructions

1. Create custom block with `ComboboxControl`
2. Start typing to have 2+ options to be shown.
3. Press `DOWN`

Expected behavior: selected next option.

Actual behavior: block loses focus, options hidden.

### Screenshots, screen recording, code snippet

https://developer.wordpress.org/block-editor/reference-guides/components/combobox-control/

```js
import { ComboboxControl } from '@wordpress/components';
import { useState } from '@wordpress/element';
 
const options = [
    {
        value: 'small',
        label: 'Small',
    },
    {
        value: 'normal',
        label: 'Normal',
    },
    {
        value: 'large',
        label: 'Large',
    },
    {
        value: 'huge',
        label: 'Huge',
    },
];
 
function MyComboboxControl() {
    const [ fontSize, setFontSize ] = useState();
    const [ filteredOptions, setFilteredOptions ] = useState( options );
    return (
        <ComboboxControl
            label="Font Size"
            value={ fontSize }
            onChange={ setFontSize }
            options={ filteredOptions }
            onFilterValueChange={ ( inputValue ) =>
                setFilteredOptions(
                    options.filter( ( option ) =>
                        option.label
                            .toLowerCase()
                            .startsWith( inputValue.toLowerCase() )
                    )
                )
            }
        />
    );
}
```

Keys should be handled properly, source: https://github.com/WordPress/gutenberg/blob/39f883e503f05aac223fe9d9d03a49d6279192a4/packages/components/src/combobox-control/index.js#L107-L141

### Environment info

- WordPress 5.9
- Chrome 97.0.4692.99
- macOS 12.2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by mahnunchik** [comment]
Key `UP` works as expected.

**Comment by t-hamano** [comment]
I have tested it in a similar environment and confirmed that the same symptom occurs with `UP` key.

The option does not change and it loses focus **if there is a block or title element before / after the `ComboboxControl`.**

For some reason, I think the keyboard events are not working.

Please see:

https://user-images.githubusercontent.com/54422211/152681534-0d7a5b55-4dae-4bb3-9bea-40ea15f52ca8.mp4

- 00:00～: There is a block before ComboboxControl.
- 00:17～: There is a block after ComboboxControl.
- 00:30～: There is a block before and after ComboboxControl.

### Environment info

- WordPress version: 5.9
- Gutenberg version: 12.5.3
- Chrome 97.0.4692.99 
- OS: Windows 11

**Comment by t-hamano** [comment]
I don't know the details, but I removed one of the event handlers registered in `.components-combobox-control__suggestions-container`, and it seems to work correctly.

![capture](https://user-images.githubusercontent.com/54422211/152682786-3fa326df-f5aa-439a-b5f6-94c6ead0c3f6.png)

**Comment by mahnunchik** [comment]
Just for the information, `FormTokenField` component has almost the same implementation of key handler:

https://github.com/WordPress/gutenberg/blob/5c0b338c44296688edac9945b23dbd18869b11cd/packages/components/src/form-token-field/index.js#L141-L189

I'm unable to check it now, but maybe (a) it has the same issue (b) the issue is already fixed there.

**Comment by stefanfisk** [comment]
I this caused by some parts of the code attaching the event listener directly to the DOM instead of handling the event via React?

Doing this and passing `ref` instead of `onKeyDown` to `<input>` fixes the issue for me:

```
const ref = useRefEffect((element) => {
    element.addEventListener('keydown', onInputKeyDown);

    return () => {
        element.removeEventListener('keydown', onInputKeyDown);
    };
});
```

**Comment by t-hamano** [comment]
I am investigating, but have created code to temporarily enable the block so that anyone can reproduce this issue.
If you run this code in your browser console, a custom block will be added and you should be able to reproduce this issue.

<details>
<summary>Code</summary>

```javascript
( function ( blocks, components, element, blockEditor ) {
	const el = element.createElement;
	blocks.registerBlockType( 'create-block/test-block', {
		title: '#38548 Test Block',
		apiVersion: 2,
		icon: 'smiley',
		category: 'text',
		edit: function () {
			return el(
				'div',
				blockEditor.useBlockProps({
					style: {
						backgroundColor: '#eee',
						padding: '16px',
						border: '1px solid #aaa',
					}
				}),
				el( components.ComboboxControl, {
					label: 'Combobox Control',
					options: [
						{
							value: 'small',
							label: 'Small',
						},
						{
							value: 'normal',
							label: 'Normal',
						},
						{
							value: 'large',
							label: 'Large',
						},
						{
							value: 'huge',
							label: 'Huge',
						},
					],
				} )
			);
		},
		save: function () {
			return null;
		},
	} );
} )(
	window.wp.blocks,
	window.wp.components,
	window.wp.element,
	window.wp.blockEditor
);
```
</details>

https://user-images.githubusercontent.com/54422211/190146520-0f3c649d-9405-4635-a5ee-d214eede7268.mp4

**Comment by t-hamano** [comment]
I have not been able to solve this problem, but I will describe what I have learned from my research so far in the hope that someone else will be able to solve this problem.

First, in the `ComboboxControl` component, the keyboard operations are tracked here:

https://github.com/WordPress/gutenberg/blob/eeeccf61f91ee64bc9de74416cd66378f15fd42f/packages/components/src/combobox-control/index.js#L121-L128

If there is a focusable element (block or title) above or below, starting from the block containing the `ComboboxControl` component, `event.defaultPrevented` will be `true` and no movement within the dropdown will occur.

What occurs instead is the vertical movement of the caret (i.e., the movement of the block in focus) in the `useArrowNav` hook:

https://github.com/WordPress/gutenberg/blob/eeeccf61f91ee64bc9de74416cd66378f15fd42f/packages/block-editor/src/components/writing-flow/use-arrow-nav.js#L263-L270

If we were to comment out the `placeCaretAtVerticalEdge`, the cursor movement within the dropdown would work correctly.

However, in the `ComboboxControl` component, if the cursor up/down operation is caught, `event.preventDefault();` should be performed here:

https://github.com/WordPress/gutenberg/blob/eeeccf61f91ee64bc9de74416cd66378f15fd42f/packages/components/src/combobox-control/index.js#L135-L154

And in the `useArrowNav` hook, the processing should be canceled at this point:

https://github.com/WordPress/gutenberg/blob/eeeccf61f91ee64bc9de74416cd66378f15fd42f/packages/block-editor/src/components/writing-flow/use-arrow-nav.js#L221-L224

However, for some reason `event.defaultPrevented` is always `false`, so the vertical movement of focus to the block takes precedence.

I think we need to add the appropriate treatment to either `ComboboxControl` or `useArrowNav`, but for the life of me, I could not pinpoint the cause.

I have given it a hoge label and hope someone familiar with this issue can solve it.

**Comment by stefanfisk** [comment]
@t-hamano did you look into this https://github.com/WordPress/gutenberg/issues/38548#issuecomment-1110924135?

From what I can tell, event handlers that are registered this way take precedence over those registered via React, which makes the event bubbling super confusing.

**Comment by t-hamano** [comment]
So, if we register an event handler via React in `ComboboxControl` as well, it will solve this problem?
If so, I would like to try to fix it.
Of course, the PR from you would be greatly appreciated 👍

**Comment by stefanfisk** [comment]
If I remember correctly, and assuming that my findings were correct, the problem is not in `<ComboboxControl>`. Instead it is `useArrowNav()` (or some other similar hook) which registers its `keydown` handler via `addEventListener()`, and therefore handles the event before `<ComboboxControl>` has had a chance to call `event.preventDefault()`.

So a quick and dirty fix is to make `<ComboboxControl>` also use `addEventListener()`. But IMHO that's just kicking the can down the road, what is really needed is to stop using `addEventListener` an instead let React handle all events.

This was all a few months back though, and right now I've got a two month old baby taking up all of my time, so I don't have the time to confirm any of this.

**Comment by stefanfisk** [comment]
But if you find the code that changes the selection and set a breakpoint there as well as in `<ComboxboxControl>`s `onKeyDown`, you should be able to verify if this is the root cause.

**Comment by t-hamano** [comment]
Thanks for the informative info!
Since `addEventListener` is used everywhere, I don't know if it is possible to completely replace them.
Therefore, I would like to create a PR using the approach you suggested and ask other members for their input.

**Comment by stefanfisk** [comment]
Yeah, that's such a fundamental architectural change that I just backed away slowly once I got that far 😕

**Comment by t-hamano** [comment]
I have confirmed that the approach described in this comment solves this problem on the block editor.

However, there is no longer a proper way to "escape" from the `ComboboxControl` on the block editor.
The up/down keys change the choices. The left and right keys move the cursor in the text box. In other words, the arrow keys alone will not allow us to escape from the `ComboboxControl` and focus on other blocks. The only way is to press the tab key to give focus once to the reset button and then press the up/down keys.

I believe this is inappropriate in terms of accessibility and would like to explore what keystrokes would be appropriate.

**Comment by mirka** [comment]
> However, there is no longer a proper way to "escape" from the `ComboboxControl` on the block editor.
> The up/down keys change the choices. The left and right keys move the cursor in the text box. In other words, the arrow keys alone will not allow us to escape from the `ComboboxControl` and focus on other blocks. The only way is to press the tab key to give focus once to the reset button and then press the up/down keys.

There are already cases where the special arrow key handling of the Writing Flow is just not compatible with a control's own standard arrow key handling. For example, a standard `<input type="number">` or `<input type="time">` where the up/down keys are supposed to change the value of the input. Edge navigation is [already disabled](https://github.com/WordPress/gutenberg/blob/000872e685e352e411b0958af9aef5da4d477532/packages/block-editor/src/components/writing-flow/use-arrow-nav.js#L33) in some of those known cases, requiring a <kbd>Tab</kbd> to navigate out.

I would argue that we should first honor the standard arrow controls of the underlying component ([combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) in this case), rather than prioritize the Writing Flow edge navigation, as this is more of an enhancement on top of the standard controls.

In that vein, one possible way to address might be to add `role="listbox"` to the opt-outs in `isNavigationCandidate()`.

**Comment by talldan** [comment]
Put together a PR to fix this in #74089 using the suggestion of checking the role.
